sdk/src/azure/core/az_mqtt5_rpc_client_hfsm.c

```diff
@@ -0,0 +1,691 @@
+      break;
+
+    case AZ_HFSM_EVENT_EXIT:
+      _rpc_stop_timer(this_policy);
+      this_policy->_internal.pending_subscription_id = 0;
+      break;
+
+    case AZ_MQTT5_EVENT_SUBACK_RSP:
+    {
+      // if get suback that matches the sub we sent, stop waiting for the suback
+      az_mqtt5_suback_data* data = (az_mqtt5_suback_data*)event.data;
+      if (data->id == this_policy->_internal.pending_subscription_id)
+      {
+        _az_RETURN_IF_FAILED(_az_hfsm_transition_peer((_az_hfsm*)me, subscribing, ready));
+      }
+      break;
+    }
+
+    case AZ_HFSM_EVENT_TIMEOUT:
+    {
+      if (event.data == &this_policy->_internal.rpc_client_timer)
+      {
+        // if subscribing times out, go to faulted state - this is not recoverable
+        _az_RETURN_IF_FAILED(_az_hfsm_transition_peer((_az_hfsm*)me, subscribing, faulted));
+      }
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_PUB_RECV_IND:
+    {
+      az_mqtt5_recv_data* recv_data = (az_mqtt5_recv_data*)event.data;
+
+      // Ensure pub is of the right topic
+      if (az_span_is_content_equal(
+              this_policy->_internal.rpc_client->response_topic, recv_data->topic))
+      {
+        // transition to ready - we must already be subscribed
+        _az_RETURN_IF_FAILED(_az_hfsm_transition_peer((_az_hfsm*)me, subscribing, ready));
+
+        // handle the response (same in all states, so handle in root)
+        ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      }
+      break;
+    }
+
+    case AZ_EVENT_RPC_CLIENT_UNSUB_REQ:
+    {
+      // transition to idle
+      _az_RETURN_IF_FAILED(
+          _az_hfsm_transition_peer(&this_policy->_internal.rpc_client_policy, subscribing, idle));
+
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+    }
+
+    case AZ_EVENT_RPC_CLIENT_INVOKE_REQ:
+    {
+      return AZ_ERROR_HFSM_INVALID_STATE;
+      break;
+    }
+
+    case AZ_EVENT_RPC_CLIENT_SUB_REQ:
+      // already subscribing, application doesn't need to take any action
+      ret = AZ_OK;
+      break;
+
+    default:
+      // TODO
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+static az_result ready(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+  az_mqtt5_rpc_client_hfsm* this_policy = (az_mqtt5_rpc_client_hfsm*)me;
+
+  if (_az_LOG_SHOULD_WRITE(event.type))
+  {
+    _az_LOG_WRITE(event.type, AZ_SPAN_FROM_STR("az_rpc_client_hfsm/ready"));
+  }
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+      // notify application that it can now send invoke requests
+      // if ((az_event_policy*)this_policy->inbound_policy != NULL)
+      // {
+      // az_event_policy_send_inbound_event((az_event_policy*)this_policy, (az_event){.type =
+      // AZ_EVENT_RPC_SERVER_EXECUTE_COMMAND_REQ, .data = data});
+      // }
+      // TODO: figure out data type here that allows app to know what rpc client sent this
+      _az_RETURN_IF_FAILED(_az_mqtt5_connection_api_callback(
```

Comment:
Consider moving the callback as an action of the transition, unless it is required in all cases when this state becomes active.

Usually, ENTRY/EXIT events contain memory allocation/deallocation, initialization/deallocation, etc and not actions (such as run the callback code) - see 2.3.6 from https://www.state-machine.com/doc/PSiCC2.pdf.
That said, if we need to guarantee the callback is called every time this state is entered, the code is correct.

Reply:
For entering the Faulted state, I think it definitely makes sense to always trigger sending the error event, so I'll leave that one. For this case, I can see it going either way - we always want to send this event when we hit the ready state (regardless of whether it's through a suback or a pub recv), but I can move it out to those transitions to be more future proof